cmd/consumer/main.go

```diff
@@ -0,0 +1,256 @@
+	WorkerCount            int
+	WorkerIntervalDuration time.Duration
+	DatabaseDSN            string
+}
+
+func loadConfig() (*Config, error) {
+	viper.SetConfigFile(".env")
+	if err := viper.ReadInConfig(); err != nil {
+		return nil, errors.Wrap(err, "failed to read env file")
+	}
+
+	c := &Config{
+		WorkerCount:            runtime.NumCPU(),
+		WorkerIntervalDuration: 300 * time.Second,
+		DatabaseDSN: fmt.Sprintf(
+			"postgres://%s:%s@%s:%s/%s",
+			viper.GetString("DATABASE_USER"),
+			viper.GetString("DATABASE_PASSWORD"),
+			viper.GetString("DATABASE_HOST"),
+			viper.GetString("DATABASE_PORT"),
+			viper.GetString("DATABASE_DB"),
+		),
+	}
+
+	intervalSeconds := viper.GetInt("WORKER_INTERVAL_SECONDS")
+	if intervalSeconds != 0 {
+		c.WorkerIntervalDuration = time.Duration(intervalSeconds) * time.Second
+	}
+
+	return c, nil
+}
+
+func main() {
+	if err := run(); err != nil {
+		log.Fatal(err)
+		os.Exit(1)
+	}
+}
+
+func run() error {
+	ctx, cancelFn := context.WithCancel(context.Background())
+	defer cancelFn()
+
+	go func() {
+		// handle interrupts and propagate the changes across the consumer pipeline
+		c := make(chan os.Signal, 1)
+		signal.Notify(c, os.Interrupt)
+		<-c
+		cancelFn()
+	}()
+
+	cfg, err := loadConfig()
+	if err != nil {
+		return errors.Wrap(err, "loading config")
+	}
+
+	logger, err := zap.NewProduction()
+	if err != nil {
+		return errors.Wrap(err, "creating logger")
+	}
+	defer logger.Sync()
+
+	logger.Info("opening database connection")
+	db, err := database.New(ctx, cfg.DatabaseDSN)
+	if err != nil {
+		return errors.Wrap(err, "opening store db connection")
+	}
+	defer db.Close()
+
+	logger.Info("creating HN client")
+	hn := hn.New()
+
+	consumer := &Consumer{
+		logger:      logger,
+		db:          db,
+		hn:          hn,
+		workerCount: cfg.WorkerCount,
+	}
+
+	for {
+		logger.Info("running consumer")
+		err := consumer.run(ctx)
+		if err != nil {
+			logger.Error("consumer failed", zap.Error(err))
+		}
+
+		select {
+		case <-ctx.Done():
+			return nil
+		case <-time.Tick(cfg.WorkerIntervalDuration):
+		}
+	}
+}
+
+type DBWriter interface {
+	Insert(ctx context.Context, item database.Item) error
+}
+
+type Consumer struct {
+	logger      *zap.Logger
+	db          DBWriter
+	hn          *hn.Client
+	workerCount int
+}
+
+// run fetches the top story ids from Hacker News and passes them along a pipeline to be fetched and inserted
+func (w *Consumer) run(ctx context.Context) error {
+	ids, err := w.hn.FetchTopStories()
+	if err != nil {
+		return errors.Wrap(err, "fetching top stories")
+	}
+	w.logger.Info("fetched top story ids", zap.Int("count", len(ids)))
+
+	// convert the top story ids into a channel of ids
+	idStream := generator(ctx, ids)
+
+	// create workers to fetch items from the HN API
+	fetchers := make([]<-chan fetchItemResult, w.workerCount)
+	for i := 0; i < w.workerCount; i++ {
+		fetchers[i] = fetchItem(ctx, idStream, w.hn)
+	}
+
+	// fan-in the fetched items and decide what to do with them
```

Comment:
Do you need to fan in, could you insert into the database in the workers?

Reply:
It would also simplify the code

Reply:
Agreed. It's also creating a bottleneck for inserting. It was part of the Concurrency in Go book I was reading so thought I'd give it a go. Maybe it'd make more sense in other contexts